### PR TITLE
Read and propagate helm config for security products

### PR DESCRIPTION
**Pull request by robertpi**
### What does this PR do?

This PR is indent to allow ASM, IAST and SCA to be enabled (or disabled) by Helm Charts config on the controller pod. The cluster-agent, running on the controller pod, will read some environment variables and use this to mutate the configuration of other pod to set the environment variables that will activate products within the client libraries.

### Motivation

This will make it easier for k8s clients (at least those using Helm Charts) to activate our products. Simplified installation is a common request.

### Additional Notes

The PR will is designed to work with this PR: https://github.com/DataDog/helm-charts/pull/1337

Each product will have env var with the following format:

* `DD_product_ENABLED_PROPAGATE` - overall enable disable switch
* `DD_product_ENABLED_PROPAGATE_NAMESPACES`  - filter for namespaces where the products should enabled
* `DD_product_DISABLED_PROPAGATE_NAMESPACES` - filter for namespaces where the products shouldn't enabled

It will result in the a variable with the following format be propagated to all pods (or those conforming to the filters):

`DD_product_ENABLED`

### Possible Drawbacks / Trade-offs

Adds more complexity to our config handling.

### Describe how to test/QA your changes

* Unit tests have been added
* Manual end to end testing has been preformed

### CODE REVIEW

File: pkg/config/setup/apm.go [hunk 1]
```diff
@@ -82,6 +82,16 @@ func setupAPM(config pkgconfigmodel.Config) {
...
+	config.BindEnv("asm_config.enabled", "DD_APPSEC_ENABLED_PROPAGATE")
+	config.BindEnvAndSetDefault("asm_config.enabled_namespaces", []string{}, "DD_APPSEC_ENABLED_PROPAGATE_NAMESPACES")
+	config.BindEnvAndSetDefault("asm_config.disabled_namespaces", []string{}, "DD_APPSEC_DISABLED_PROPAGATE_NAMESPACES")
+	config.BindEnv("iast_config.enabled", "DD_IAST_ENABLED_PROPAGATE")
+	config.BindEnvAndSetDefault("iast_config.enabled_namespaces", []string{}, "DD_IAST_ENABLED_PROPAGATE_NAMESPACES")
+	config.BindEnvAndSetDefault("iast_config.disabled_namespaces", []string{}, "DD_IAST_DISABLED_PROPAGATE_NAMESPACES")
+	config.BindEnv("sca_config.enabled", "DD_SCA_ENABLED_PROPAGATE")
+	config.BindEnvAndSetDefault("sca_config.enabled_namespaces", []string{}, "DD_SCA_ENABLED_PROPAGATE_NAMESPACES")
+	config.BindEnvAndSetDefault("sca_config.disabled_namespaces", []string{}, "DD_SCA_DISABLED_PROPAGATE_NAMESPACES")
+
```
Comment (adel121) [inline pkg/config/setup/apm.go:94]: ❓ 

I wonder why you placed these configurations here unde`apm.go` 
AFAIK, these three features are not related to APM, should these configs be places under `pkg/config/setup/config.go` ?

Comment (adel121) [inline pkg/config/setup/apm.go:94]: I don't think you need to use `asm_config.*`, `asm.*` should be enough. (same for all the others)

File: pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go [hunk 2]
```diff
@@ -141,17 +141,39 @@ type Webhook struct {
 	endpoint          string
 	resources         []string
 	operations        []admiv1.OperationType
-	filter            *containers.Filter
+	filters           map[string]*containers.Filter
```
Comment (adel121) [inline pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:144]: I don't think this is a good idea. 

This leads to coupling within the same webhook multiple features that are not actually related.

We should probably have one webhook per feature (or a webhook for a set of features only if it makes sense to, which I assume is a very rare situation)

File: pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go [hunk 3]
```diff
@@ -141,17 +141,39 @@ type Webhook struct {
...
+
+	var filters map[string]*containers.Filter = make(map[string]*containers.Filter);
+
+	ssiFilter, err := apmSSINamespaceFilter();
+	if err != nil {
+		return nil, err
+	}
+	filters["apm_config.instrumentation.enabled"] = ssiFilter;
+
+	asmFilter, err := apmProductNamespaceFilter("asm_config.enabled_namespaces", "asm_config.disabled_namespaces");
```
Comment (adel121) [inline pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:160]: ASM is not related to APM afaik (correct me if I am wrong).

Given this, I don't think the naming is good here (i.e. `apmProductNamespaceFilter(...)` is returning an `asm` filter)

Same applies to the following lines.

File: pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go [hunk 4]
```diff
@@ -141,17 +141,39 @@ type Webhook struct {
...
 	}
+	filters["asm_config.enabled"] = asmFilter;
+
+	iastFilter, err := apmProductNamespaceFilter("iast_config.enabled_namespaces", "iast_config.disabled_namespaces");
+	if err != nil {
+		return nil, err
+	}
+	filters["iast_config.enabled"] = iastFilter;
+
+	scaFilter, err := apmProductNamespaceFilter("sca_config.enabled_namespaces", "sca_config.disabled_namespaces");
```
Comment (adel121) [inline pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:172]: same comment as above

File: pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go [hunk 5]
```diff
@@ -141,17 +141,39 @@ type Webhook struct {
...
 	if err != nil {
 		return nil, err
 	}
+	filters["asm_config.enabled"] = asmFilter;
+
+	iastFilter, err := apmProductNamespaceFilter("iast_config.enabled_namespaces", "iast_config.disabled_namespaces");
+	if err != nil {
+		return nil, err
+	}
+	filters["iast_config.enabled"] = iastFilter;
```
Comment (adel121) [inline pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:170]: I don't think it is good to have the absolute config name as a key in the filters map. (same applies to the rest).

Using filters["iast"] should be enough, and would be better to create some constant like `const iast = "iast"` and use it wherever needed; this helps avoid runtime errors in case of typo in the key when used as a plain string.

But despite this:
- It is not good to have all the features in the same webhook in the first place

File: pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go [hunk 6]
```diff
@@ -141,17 +141,39 @@ type Webhook struct {
...
+
+	scaFilter, err := apmProductNamespaceFilter("sca_config.enabled_namespaces", "sca_config.disabled_namespaces");
+	if err != nil {
+		return nil, err
+	}
+	filters["sca_config.enabled"] = scaFilter;
```
Comment (eliottness) [inline pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:176]:
```suggestion
    var err error
	filters["apm_config.instrumentation.enabled"], err = apmSSINamespaceFilter()
	if err != nil {
		return nil, err
	}

	filters["asm_config.enabled"], err = apmProductNamespaceFilter("asm_config.enabled_namespaces", "asm_config.disabled_namespaces")
	if err != nil {
		return nil, err
	}

	filters["iast_config.enabled"], err = apmProductNamespaceFilter("iast_config.enabled_namespaces", "iast_config.disabled_namespaces")
	if err != nil {
		return nil, err
	}

	filters["sca_config.enabled"], err = apmProductNamespaceFilter("sca_config.enabled_namespaces", "sca_config.disabled_namespaces")
	if err != nil {
		return nil, err
	}
```

File: pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go [hunk 7]
```diff
@@ -552,20 +584,43 @@ func ShouldInject(pod *corev1.Pod) bool {
...
-	return apmWebhook.isEnabledInNamespace(pod.Namespace) || config.Datadog.GetBool("admission_controller.mutate_unlabelled")
+	return apmWebhook.isSsiEnabledInNamespace(pod.Namespace) || config.Datadog.GetBool("admission_controller.mutate_unlabelled")
 }
 
-// isEnabledInNamespace indicates if Single Step Instrumentation is enabled for
+// isEnabledInNamespace indicates if a product is enabled for the namespace in the cluster
+func (w *Webhook) isEnabledInNamespace(enabledConfig string, namespace string) bool {
+	enabledValue := config.Datadog.GetString(enabledConfig)
+
+	if len(enabledValue) == 0 {
```
Comment (eliottness) [inline pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:594]:
```suggestion
	if enabledValue == ""  {
```

File: pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go [hunk 8]
```diff
@@ -552,20 +584,43 @@ func ShouldInject(pod *corev1.Pod) bool {
...
+
+func (w *Webhook) injectProductActivation(pod *corev1.Pod, enableKey string, envVarKey string) error {
+	if (w.isEnabledInNamespace(enableKey, pod.Namespace)) {
+		enabledValue := config.Datadog.GetBool(enableKey)
+		_ = mutatecommon.InjectEnv(pod, corev1.EnvVar{
+				Name:	envVarKey,
+				Value: 	strconv.FormatBool(enabledValue),
+			});
+	}
+	return nil;
```
Comment (eliottness) [inline pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:623]:
```suggestion
	if w.isEnabledInNamespace(enableKey, pod.Namespace) {
		enabledValue := config.Datadog.GetBool(enableKey)
		mutatecommon.InjectEnv(pod, corev1.EnvVar{
				Name:	envVarKey,
				Value: 	strconv.FormatBool(enabledValue),
			})
	}
	return nil
```

File: pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go [hunk 9]
```diff
@@ -229,6 +252,11 @@ func apmSSINamespaceFilter() (*containers.Filter, error) {
 	return containers.NewFilter(containers.GlobalFilter, apmEnabledNamespacesWithPrefix, filterExcludeList)
 }
 
+// apmSSINamespaceFilter returns the filter used by APM SSI to filter namespaces.
+func apmSSINamespaceFilter() (*containers.Filter, error) {
+	return apmProductNamespaceFilter("apm_config.instrumentation.enabled_namespaces", "apm_config.instrumentation.disabled_namespaces");
```
Comment (eliottness) [inline pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go:257]:
```suggestion
	return apmProductNamespaceFilter("apm_config.instrumentation.enabled_namespaces", "apm_config.instrumentation.disabled_namespaces")
```

**Review by eliottness — APPROVED** [review]
Couple of go nitpicks but thanks @robertpi :100:

**Review by eliottness — COMMENTED** [review]
Woops, missclicked on the approve button
